graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -967,45 +1007,16 @@ def _compile_root_ast_to_ir(schema, ast, type_equivalence_hints=None):
     # Based on the outputs context data, add an output step.
     basic_blocks.append(_compile_output_step(query_metadata_table))
 
-    # Construct the output metadata, ensuring that all folded outputs have a valid type
-    # (GraphQLInt for _x_count, GraphQLList for all other folded outputs) and that _x_count is only
-    # output within a fold scope.
-    output_metadata = {}
-    for name, info in query_metadata_table.outputs:
-        # Ensure _x_count is within a fold scope.
-        if info.location.field == COUNT_META_FIELD_NAME and not isinstance(
-            info.location, FoldScopeLocation
-        ):
-            raise AssertionError(
-                u"Invalid output: {} was not in a fold scope.".format(COUNT_META_FIELD_NAME)
-            )
-        # Ensure folded outputs have valid type.
-        if isinstance(info.location, FoldScopeLocation):
-            # Ensure _x_count is a GraphQLInt.
-            if info.location.field == COUNT_META_FIELD_NAME and not is_same_type(
-                info.type, GraphQLInt
-            ):
-                raise AssertionError(
-                    u"Invalid output: received {} with type {}, but {} must always be of "
-                    u"type GraphQLInt".format(
-                        COUNT_META_FIELD_NAME, info.type, COUNT_META_FIELD_NAME
-                    )
-                )
-            # Ensure all other folded outputs are GraphQLList.
-            elif info.location.field != COUNT_META_FIELD_NAME and not isinstance(
-                info.type, GraphQLList
-            ):
-                raise AssertionError(
-                    u"Invalid output: non-{} folded output must have type "
-                    u"GraphQLList. Received type {} for folded output "
-                    u"{}.".format(COUNT_META_FIELD_NAME, info.type, name)
-                )
-        # Construct output metadata.
-        output_metadata[name] = OutputMetadata(
+    # Construct the output metadata, ensuring that all folded outputs have a valid type.
+    _validate_output_types(query_metadata_table.outputs)
+    output_metadata = {
+        name: OutputMetadata(
```

Comment:
An alternative setup might be something like:
```
output_metadata = {
    name: make_output_metadata(name, info)
    for name, info in query_metadata_table.outputs
}
```
with all of the validation and output metadata creation in the `make_output_metadata` function. That might make it a bit easier to read and follow along.